coalib/core/Core.py

```diff
@@ -327,34 +150,177 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
     return dependency_tracker, bears
 
 
-def run(bears, result_callback):
+class Session:
     """
-    Runs a coala session.
+    Maintains Session for a coala execution. For each session, there are set
+    of bears along with callback function, which is run when results are
+    available. On running a session, first its dependencies are initalized and
+    all the bears are scheduled to run in appropriate order.
```

Comment:
a bit grammatical polishing ;)
-->
```
Maintains a session for a coala execution. For each session, there are set bears to run along with a callback function, which is run when results are available.
```
About this part:
```
On running a session, first its dependencies are initalized and
all the bears are scheduled to run in appropriate order.
```
I would remove that altogether, as 1. it's implementation detail 2. it's expected from the user that this works, so imo it requires no mention. And `all the bears are scheduled to run in appropriate order` is technically not completely correct ;) Bears that depend on another bear aren't scheduled until the bear dependency has finished. So in the beginning not all bears may be scheduled :)